app/services/segment_delete_service.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+class SegmentDeleteService
+  SEGMENT_URI = URI('https://gdpr.segment.com/graphql')
+
+  class SegmentDeleteServiceError < StandardError
+    def initialize(message:, response:)
+      super(message)
+      @response = {body: response.body, code: response.code}
+    end
+
+    attr_reader :response
+
+    def error_type
+      :delete_service
+    end
+  end
+  class SegmentAuthenticationError < SegmentDeleteServiceError
+    def initialize(message: 'Cannot authenticate', response:)
+      super(message: message, response: response)
+    end
+
+    def error_type
+      :authentication
+    end
+  end
+  class SegmentUserDeletionError < SegmentDeleteServiceError
+    def initialize(message: 'Cannot delete user', response:)
+      super(message: message, response: response)
+    end
+
+    def error_type
+      :user_deletion
+    end
+  end
+
+
+  def initialize(user_id)
+    @user_id = user_id
+    @errors = ActiveModel::Errors.new(self)
+  end
+
+  attr_reader :user_id, :errors
+
+  def call!
+    return true unless Features::SegmentDeletionConfig.enabled?
+    request_body = {query: "mutation { createWorkspaceRegulation(workspaceSlug: \"3scale\" type: SUPPRESS_AND_DELETE userId: \"#{user_id}\" ) { id } }"}.to_json
+    response = perform_request(request_body: request_body, authorization: request_token)
+    raise SegmentUserDeletionError.new(response: response) unless response.code.to_s == '200'
+    :ok
+  end
+
+  def call
+    call!
+  rescue SegmentDeleteServiceError => exception
+    errors.add(exception.error_type, exception.response)
+    false
+  end
+
+  def request_token
+    return true unless Features::SegmentDeletionConfig.enabled?
+    request_body = {
+      query: 'mutation auth($email:String!, $password:String!) {login(email:$email, password:$password)}',
+      variables: Features::SegmentDeletionConfig.config
+    }.to_json
+    response = perform_request(request_body: request_body)
+    authentication_json_response(response: response).dig('data', 'login', 'access_token') or raise SegmentAuthenticationError.new(response: response)
+  end
+
+  private
+
+  def authentication_json_response(response:)
+    JSON.parse(response.body)
+  rescue JSON::ParserError
+    raise SegmentAuthenticationError.new(response: response)
+  end
+
+  def perform_request(request_body:, authorization: nil)
+    headers = {'Content-Type' => 'application/json; charset=utf-8'}
+    headers['Authorization'] = "Bearer #{authorization}" if authorization
+    request = Net::HTTP::Post.new(SEGMENT_URI.path, headers)
```

Comment:
OK you can reuse libraries already installed in the project. Though Standard library is not bad for that. But lets not use too many different implementations.

You have the choice, though they do not have the same level of purpose:

- Faraday
- RestClient
- HTTPClient

Reply:
Recommendation from Hery for this case: **Faraday**